Remove vec from fluxfunction and fix compilation

ykempf opened the pull request:
This includes #1026 at the moment, apologies.

Should address #940.

Adds fluxfunction to the tools built in github CI even though we don't use it (yet). That binary is not uploaded, either.

Fixes compilation of fluxfunction and removes the vectorclass from there, and the couple files it uses from particles/. Otherwise in particles/ only converting Vec3d to Vec3Dd so it compiles.

Neither fluxfunction nor particle pusher are tested so far!

ykempf committed:
Build fluxfunction in github CI too.

ykempf committed:
Fix compilation of fluxfunction, remove vectorclass form there to use std::vector as well as the files in partilces/ used by fluxfunction.

ykempf committed:
Vec3d -> Vec3Dd in particles/.

ursg commented on particles/field.h:
```diff
@@ -60,79 +58,157 @@ struct Field
    }
 
-   Vec3d getCell(int x, int y, int z) {
+   std::vector<double> getCell(int x, int y, int z) {
```
I would suggest that this be ```std::array<double, 3>```. Because why would you use a variable-sized vector here?

ursg reviewed (changes requested):
Well, hum. Functional thumbs up, because a working fluxfunction is obviously better than a broken one.

But instead of partially keeping with vectorclass and partially unrolling vector operations into handcrafted scalar operations, maybe the better way to go here would be to just switch to Eigen vectors, like in other parts of the code (like the ionosphere)?

Also, ```std::vector<double>``` as a vector type instead of ```std::array<double,3>``` really doesn't look right to me.

ursg committed:
Change std::vectors to std::array<double,3> in fluxfunction

ykempf committed:
Merge pull request #15 from ursg/remove_vec_from_fluxfunction_and_fix_compilation

Change std::vectors to std::array<double,3> in fluxfunction

markusbattarbee commented:
+1 for using Eigen vectors instead.

ursg committed:
Replace vectorclass with Eigen in particle pusher and fluxfunction

ursg commented:
[This](https://github.com/ykempf/vlasiator/pull/16) Pull request against this PR introduces Eigen vectors instead.

ursg committed:
Actually test flux function as part of CI

this just runs on a random file of some old run lying around on
carrington, and should give the same result there.

ykempf commented on particles/boundaries.h:
```diff
@@ -140,7 +140,7 @@ struct ReflectBoundary : public Boundary
    // Vector to multiply with in order to flip velocity
    // vectors for our dimension
-   Vec3d flip_v;
+   Vec3Dd flip_v;
```
But so we still want to keep these as `vectorclass`, the others as `Eigen` vectors?

ykempf replied:
Uh wait that's old.

ykempf replied:
But still a valid question from me.

ursg committed:
Fix typo in CI Yaml (flixfunction)

ursg committed:
Include fluxfunction test in testpackage. Also generate reference data.

ykempf committed:
Merge pull request #16 from ursg/eigenParticlesFluxfunction

Replace vectorclass with Eigen in particle pusher and fluxfunction

ursg committed:
Actually run the fluxfunction test on a carrington compute node.

ursg committed:
Only link the required libraries into fluxfunction.

ursg committed:
Also don't unnecessarily link -lgfortran on carrington anymore.

(This was making the fluxfunction test fail.)

ursg committed:
Build tools on carrington with newer GCC & modules

ursg committed:
Merge remote-tracking branch 'origin/dev' into remove_vec_from_fluxfunction_and_fix_compilation

ursg committed:
Merge remote-tracking branch 'origin/dev' into remove_vec_from_fluxfunction_and_fix_compilation

ursg committed:
Run fluxfunction test with the same gcc modules as they are built with

ursg committed:
Actually srun fluxfunction test with bash -c around.

ykempf committed:
Merge branch 'dev' into remove_vec_from_fluxfunction_and_fix_compilation

ykempf committed:
Change path to find files for fluxfunction test.

ykempf commented on .github/workflows/github-ci.yml:
```diff
@@ -466,6 +469,13 @@ jobs:
+        chmod +x $GITHUB_WORKSPACE/fluxfunction
+        srun --job-name fluxtest -M carrington -N 1 -c 1 --mem=1G bash -c "module purge; module load GCC/13.2.0; module load OpenMPI/4.1.6-GCC-13.2.0 ; module load PMIx/4.2.6-GCCcore-13.2.0; module load PAPI/7.1.0-GCCcore-13.2.0; $GITHUB_WORKSPACE/fluxfunction testpackage/run_*/Magnetosphere_small/bulk.0000001.vlsv equatorial.bin"
+        srun --job-name fluxtest -M carrington -N 1 -c 1 --mem=1G bash -c "module purge; module load GCC/13.2.0; module load OpenMPI/4.1.6-GCC-13.2.0 ; module load PMIx/4.2.6-GCCcore-13.2.0; module load PAPI/7.1.0-GCCcore-13.2.0; $GITHUB_WORKSPACE/fluxfunction testpackage/run_*/Magnetosphere_polar_small/bulk.0000001.vlsv polar.bin"
+        diff -q equatorial.bin /proj/group/spacephysics/vlasiator_testpackage/CI_reference/equatorial.bin
```
OK so this now reported nothing as expected, but if it were saying there's a diff it wouldn't be very prominent. How do we raise that?

markusbattarbee replied:
pipe the output into a variable, and if the variable is acceptable, touch a file? then can have a yaml action checking for that file and failing on file not found?
@ursg might have a more elegant solution.

markusbattarbee commented on .github/workflows/github-ci.yml:
```diff
@@ -430,6 +431,8 @@ jobs:
         chmod +x $GITHUB_WORKSPACE/vlsv*_DP
         cd testpackage
+        # Delete any old run directories
+        rm -rf run_20*
```
This will break in 75 years time! ;)

markusbattarbee commented on particles/scenario.cpp:
```diff
@@ -22,6 +22,12 @@
+#define cross_product(av,bv) (av).cross(bv)
+#define dot_product(av,bv) (av).dot(bv)
+#define vector_length(v) (v).norm()
+#define normalize_vector(v) (v).normalized()
```
These are already defined in relativistic_math.h

ursg replied:
I simply used these defines as a drop-in replacement for vectorclass. I originally took them from sysboundaries/ionosphere.cpp

So... maybe a global Eigen-as-vectorclass-header would make sense?

markusbattarbee commented on particles/relativistic_math.h:
```diff
@@ -20,9 +20,14 @@
+#define dot_product(av,bv) (av).dot(bv)
+#define vector_length(v) (v).norm()
+#define normalize_vector(v) (v).normalized()
+
```
place in fields.h instead?

markusbattarbee commented on particles/particle_post_pusher.cpp:
```diff
@@ -37,12 +37,17 @@
+#define dot_product(av,bv) (av).dot(bv)
+#define vector_length(v) (v).norm()
+#define normalize_vector(v) (v).normalized()
+
```
already defined in relativistic_math.h

markusbattarbee commented on particles/histogram.h:
```diff
@@ -185,7 +187,7 @@ class LinearHistogram2D : public Histogram2D
       virtual void addValue(Vec2d value, double weight=1.) {
          value -= low;
-         value /= high - low;
+         value = value.cwiseQuotient(high - low);
```
should this also be a macro like the other vector operations?

ursg committed:
Merge remote-tracking branch 'origin/dev' into remove_vec_from_fluxfunction_and_fix_compilation

ursg committed:
Merge remote-tracking branch 'ykempf/remove_vec_from_fluxfunction_and_fix_compilation' into remove_vec_from_fluxfunction_and_fix_compilation
